### PR TITLE
Speed up parameter resolution by checking if val is parameterized and caching the boolean

### PR DESCRIPTION
**Pull request by tanujkhattar**
Fixes https://github.com/quantumlib/Cirq/issues/6001

cc @maffoo @zchen088 PTAL!

### CODE REVIEW

File: cirq-core/cirq/protocols/resolve_parameters_test.py [hunk 1]
```diff
@@ -35,16 +35,16 @@ def __init__(self, var):
             self.parameter = var
 
         def _is_parameterized_(self) -> bool:
-            return self.parameter == 0
+            return self.parameter != 0
```
Comment (zchen088) [inline cirq-core/cirq/protocols/resolve_parameters_test.py:38]: Just curious, why this change?

**Review by zchen088 — APPROVED** [review]
LGTM

File: cirq-core/cirq/ops/pauli_sum_exponential.py [hunk 2]
```diff
@@ -78,6 +78,9 @@ def _value_equality_values_(self) -> Any:
     def with_qubits(self, *new_qubits: 'cirq.Qid') -> 'PauliSumExponential':
         return PauliSumExponential(self._pauli_sum.with_qubits(*new_qubits), self._exponent)
 
+    def _is_parameterized_(self) -> bool:
```
Comment (maffoo) [inline cirq-core/cirq/ops/pauli_sum_exponential.py:81]: Can use `cached_method` here?

File: cirq-core/cirq/ops/raw_types.py [hunk 3]
```diff
@@ -111,6 +111,7 @@ def with_dimension(self, dimension: int) -> 'Qid':
     def _cmp_tuple(self):
         return (type(self).__name__, repr(type(self)), self._comparison_key(), self.dimension)
 
+    @cached_method
     def __hash__(self):
```
Comment (maffoo) [inline cirq-core/cirq/ops/raw_types.py:115]: nit: add return type
```suggestion
    def __hash__(self) -> int:
```

File: cirq-core/cirq/protocols/resolve_parameters.py [hunk 4]
```diff
@@ -177,6 +177,11 @@ def resolve_parameters(
     if isinstance(val, (list, tuple)):
         return cast(T, type(val)(resolve_parameters(e, param_resolver, recursive) for e in val))
 
+    getter = getattr(val, '_is_parameterized_', None)
+    result = NotImplemented if getter is None else getter()
+    if result is False:
+        return val
```
Comment (maffoo) [inline cirq-core/cirq/protocols/resolve_parameters.py:183]: nit: I'd suggest using a different var name for the function attribute in this case. Also, I'd suggest not setting `result` since that is used for the actual result of parameter resolution below, so using it for the return value of this method here feels odd
```suggestion
    is_parameterized = getattr(val, '_is_parameterized_', None)
    if is_parameterized is not None and not is_parameterized():
        return val
```

File: cirq-core/cirq/protocols/resolve_parameters_test.py (hunk 1 above)
Comment (tanujkhattar) [reply cirq-core/cirq/protocols/resolve_parameters_test.py:38]: Because in the existing test, the `_is_paramaterized_` returns `True` when the class is actually not paramaterized (i.e. `self.parameter is 0` and there's nothing to resolve). 

As a result `cirq.is_parameterized(val)` returns False when `val` is actually parameterized and therefore the output of `cirq.resolve_parameters(val)` changes. 

So this change fixes the test to make the behavior of `_is_paramaterized_` and `resolve_parameters` consistent.

File: cirq-core/cirq/protocols/resolve_parameters.py (hunk 4 above)
Comment (tanujkhattar) [reply cirq-core/cirq/protocols/resolve_parameters.py]: done. I'm comparing `is_parameterized() is False` instead to distinguish with the case when `is_parameterized()` can return `NotImplemented`.

File: cirq-core/cirq/ops/raw_types.py (hunk 3 above)
Comment (tanujkhattar) [reply cirq-core/cirq/ops/raw_types.py]: done

File: cirq-core/cirq/ops/pauli_sum_exponential.py (hunk 2 above)
Comment (tanujkhattar) [reply cirq-core/cirq/ops/pauli_sum_exponential.py:82]: done

**Review by tanujkhattar — COMMENTED** [review]
Addressed nits @maffoo 

PTAL

File: cirq-core/cirq/protocols/resolve_parameters.py [hunk 5]
```diff
@@ -177,6 +177,10 @@ def resolve_parameters(
     if isinstance(val, (list, tuple)):
         return cast(T, type(val)(resolve_parameters(e, param_resolver, recursive) for e in val))
 
+    is_parameterized = getattr(val, '_is_parameterized_', None)
+    if is_parameterized is not None and is_parameterized() is False:
```
Comment (maffoo) [inline cirq-core/cirq/protocols/resolve_parameters.py:181]: nit: Style guide says to use `not` instead of comparing to `False` (https://google.github.io/styleguide/pyguide.html#2144-decision):
```suggestion
    if is_parameterized is not None and not is_parameterized():
```
If if returns `NotImplemented` that is still considered a truthy value, so the condition will not hold and we will proceed to regular parameter resolution as before.

**Review by maffoo — APPROVED** [review]
Minor comment, then LGTM.
